OpenSim/Common/Logger.cpp

```diff
@@ -30,23 +30,93 @@
 
 using namespace OpenSim;
 
-std::shared_ptr<spdlog::logger> Logger::m_cout_logger = 
-        spdlog::stdout_color_mt("cout");
-std::shared_ptr<spdlog::sinks::basic_file_sink_mt> Logger::m_filesink = {};
-std::shared_ptr<spdlog::logger> Logger::m_default_logger;
-
-// Force creation of the Logger instane to initialize spdlog::loggers
-std::shared_ptr<OpenSim::Logger> Logger::m_osimLogger = Logger::getInstance();
-
-Logger::Logger() {
-    m_default_logger = spdlog::default_logger();
-    m_default_logger->set_level(spdlog::level::info);
-    m_default_logger->set_pattern("[%l] %v");
-    m_cout_logger->set_level(spdlog::level::info);
-    m_cout_logger->set_pattern("%v");
-    // This ensures log files are updated regularly, instead of only when the
-    // program shuts down.
+// cout logger will be initialized during static initialization time
+static std::shared_ptr<spdlog::logger> cout_logger = []() {
+    std::shared_ptr<spdlog::logger> logger = spdlog::stdout_color_mt("cout");
+    logger->set_level(spdlog::level::info);
+    logger->set_pattern("%v");
+    return logger;
+}();
+
+// default logger will be initialized during static initialization time
+static std::shared_ptr<spdlog::logger> default_logger = []() {
+    std::shared_ptr<spdlog::logger> logger = spdlog::default_logger();
+    logger->set_level(spdlog::level::info);
+    logger->set_pattern("[%l] %v");
+    return logger;
+}();
+
+// this initialization will also happen during static initialization time
+static bool other_init = []() {
     spdlog::flush_on(spdlog::level::info);
+    return true;
+}();
+
+// the file log sink (e.g. `opensim.log`) is *not* necessarily initialized
+// during static initialization time. It is only initialized when the first log
+// message is about to be written. Users *may* disable this functionality
+// before the first log message is written.
+static std::shared_ptr<spdlog::sinks::basic_file_sink_mt> m_filesink = nullptr;
+
+// if a user calls `Logger::removeFileSink` before m_filesink is initialized, a
+// flag should be set to ensure that the "first use" initialization does not
+// subsequently happen.
+static bool filesink_auto_initialization_disabled = false;
+
+// when filesink initialization *may* happen, ensure it only happens exactly
+// once globally by wrapping it in a function-local static.
+static bool _maybeInitFileLogging() {
+#ifdef OPENSIM_DISABLE_LOG_FILE
```

Comment:
Style: we never use leading underscore. The name maybeInitFileLogging is less informative, how about initFileLoggingAsNeeded, or something along these lines